xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,469 @@
+                        .build());
+      } else {
+        OrcaOobUtil.setListener(weightedSubchannel, null, null);
+      }
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    if (weightUpdateTimer != null) {
+      weightUpdateTimer.cancel();
+    }
+    super.shutdown();
+  }
+
+  private static final class WrrHelper extends ForwardingLoadBalancerHelper {
+    private final Helper delegate;
+    private final TimeProvider timeProvider;
+
+    WrrHelper(Helper helper, TimeProvider timeProvider) {
+      this.delegate = helper;
+      this.timeProvider = timeProvider;
+    }
+
+    @Override
+    protected Helper delegate() {
+      return delegate;
+    }
+
+    @Override
+    public Subchannel createSubchannel(CreateSubchannelArgs args) {
+      return new WrrSubchannel(delegate().createSubchannel(args), timeProvider);
+    }
+  }
+
+  @VisibleForTesting
+  static final class WrrSubchannel extends ForwardingSubchannel {
+    private final Subchannel delegate;
+    private final TimeProvider timeProvider;
+    private final OrcaOobReportListener oobListener = this::onLoadReport;
+    private final OrcaPerRequestReportListener perRpcListener = this::onLoadReport;
+    private volatile long lastUpdated;
+    private volatile long nonEmptySince;
+    private volatile double weight;
+    private volatile WeightedRoundRobinLoadBalancerConfig config;
+
+    WrrSubchannel(Subchannel delegate, TimeProvider timeProvider) {
+      this.delegate = checkNotNull(delegate, "delegate");
+      this.timeProvider = checkNotNull(timeProvider, "timeProvider");
+    }
+
+    private void setConfig(WeightedRoundRobinLoadBalancerConfig config) {
+      this.config = config;
+    }
+
+    @VisibleForTesting
+    void onLoadReport(MetricReport report) {
+      double newWeight = report.getCpuUtilization() == 0 ? 0 :
+              report.getQps() / report.getCpuUtilization();
+      if (newWeight == 0) {
+        return;
+      }
+      if (nonEmptySince == Integer.MAX_VALUE) {
+        nonEmptySince = timeProvider.currentTimeNanos();
+      }
+      lastUpdated = timeProvider.currentTimeNanos();
+      weight = newWeight;
+    }
+
+    @Override
+    public void start(SubchannelStateListener listener) {
+      delegate().start(new SubchannelStateListener() {
+        @Override
+        public void onSubchannelState(ConnectivityStateInfo newState) {
+          if (newState.getState().equals(ConnectivityState.READY)) {
+            nonEmptySince = Integer.MAX_VALUE;
+          }
+          listener.onSubchannelState(newState);
+        }
+      });
+    }
+
+    private double getWeight() {
+      if (config == null) {
+        return 0;
+      }
+      double now = timeProvider.currentTimeNanos();
+      if (now - lastUpdated >= config.weightExpirationPeriodNanos) {
+        nonEmptySince = Integer.MAX_VALUE;
+        return 0;
+      } else if (now - nonEmptySince < config.blackoutPeriodNanos) {
```

Comment:
We need "&& config.blackoutPeriodNanos > 0" as in C++ as nonEmptySince can be infy and blackoutPeriodNanos being non-positive should disable expiration.